app/src/androidTest/java/org/piwigo/ui/launcher/LoginTest.java

```diff
@@ -0,0 +1,126 @@
+import androidx.test.ext.junit.rules.ActivityScenarioRule;
+import androidx.test.ext.junit.runners.AndroidJUnit4;
+import androidx.test.filters.LargeTest;
+
+import static androidx.test.espresso.Espresso.onView;
+import static androidx.test.espresso.action.ViewActions.click;
+import static androidx.test.espresso.action.ViewActions.closeSoftKeyboard;
+import static androidx.test.espresso.action.ViewActions.pressImeActionButton;
+import static androidx.test.espresso.action.ViewActions.replaceText;
+import static androidx.test.espresso.action.ViewActions.scrollTo;
+import static androidx.test.espresso.assertion.ViewAssertions.doesNotExist;
+import static androidx.test.espresso.assertion.ViewAssertions.matches;
+import static androidx.test.espresso.contrib.DrawerActions.open;
+import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
+import static androidx.test.espresso.matcher.ViewMatchers.withContentDescription;
+import static androidx.test.espresso.matcher.ViewMatchers.withId;
+import static androidx.test.espresso.matcher.ViewMatchers.withText;
+import static org.hamcrest.CoreMatchers.allOf;
+
+@LargeTest
+@RunWith(AndroidJUnit4.class)
+public class LoginTest {
+
+    @Rule
+    public ActivityScenarioRule<LauncherActivity> activityScenarioRule =
+        new ActivityScenarioRule<LauncherActivity>(LauncherActivity.class);
+
+    @Test
+    public void loginTest() {
+
+        try {
+            Thread.sleep(1700);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+
+        ViewInteraction editUser = onView(withId(R.id.username));
+        editUser.check(matches(withText("")));
+
+        ViewInteraction editURL = onView(withId(R.id.url));
+        editURL.perform(replaceText("https://tg1.kulow.org"), closeSoftKeyboard());
+
+        editUser.perform(replaceText("seessmall"), closeSoftKeyboard());
+        editUser.perform(pressImeActionButton());
+
+        ViewInteraction editPassword = onView(withId(R.id.password));
+        editPassword.perform(replaceText("seessmall"), closeSoftKeyboard());
+        editPassword.perform(pressImeActionButton());
+
+        ViewInteraction loginButton = onView(allOf(withId(R.id.login_button), withContentDescription("Login")));
+        loginButton.perform(scrollTo(), click());
+
+        // Added a sleep statement to match the app's execution delay.
+        // The recommended way to handle such scenarios is to use Espresso idling resources:
+        // https://google.github.io/android-testing-support-library/docs/espresso/idling-resource/index.html
+        try {
+            Thread.sleep(5000);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+
+        ViewInteraction toolbar = onView(allOf(withText("Albums"), isDisplayed()));
+        toolbar.check(matches(isDisplayed()));
+
+        onView(withText("Small")).check(matches(isDisplayed()));
+        onView(withText("Large")).check(doesNotExist());
```

Comment:
"Small" and "Large" are Texts coming from the test gallery data, should we comment this? Assume one day the gallery on your test server is not available and someone debugs it, will he/she directly understand that it is a problem of the gallery content (after choosing another gallery URL)?

Reply:
You also need to change the login data, etc. So basically the whole test only works there. But I may add a description text before the test.

Reply:
that's true and ok. The only difference I see is that for login data I'd consider it absolutely obvious, while for the "Small" or "Large" it is not obvious that it is data from the gallery and not a text in the app.